geps/gep-1713.md

```diff
@@ -0,0 +1,103 @@
+# GEP-1713: Standard Mechanism to Merge Multiple Gateways
+
+* Issue: [#1713/kubernetes-sigs/gateway-api/issues/1713)
+* Status: Provisional
+
+(See status definitions [here](overview.md#status).)
+
+## TLDR
+
+The Gateway Resource is a contention point since it is the only place to attach listeners with certificates. We propose a mechanism to allow distinct Gateway resources to be logically merged.
+
+## Goals
+
+- Define a mechanic to merge multiple Gateways (logically)
+- Define a set of acceptable properties that can be merged and their semantics
+
+## Non-Goals
+
+- Apply a Gateway resource onto N distinct gateways (one to many)
+
+## Introduction
+
+Knative generates on demand per-service certificates using HTTP-01 challenges. There can be O(1000) Knative Services in the cluster which means we have O(1000) distinct certificates. Thus updating a single Gateway resource with this many certificates is a contention point and inhibits horizontal scaling of our controllers.
+
+The spec currently has language to indicate implemenations `MAY` merge Gateways resources but the mechanic isn't defined. 
+https://github.com/kubernetes-sigs/gateway-api/blob/541e9fc2b3c2f62915cb58dc0ee5e43e4096b3e2/apis/v1beta1/gateway_types.go#L76-L78
+
+## Future Requirement Level
+
+We want to keep this API very simple so that the merging requirement level could increase from `MAY` to `MUST`
+
+## API
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: parent-gateway
+spec:
+  gatewayClassName: example
+  listeners:
+  - name: default
+    port: 80
+    protocol: HTTP
+---
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: child-gateway
+  labels:
+    gateway.networking.k8s.io/parent-gateway: parent-gateway
+spec:
+  gatewayClassName: example
+  parentGateway: parent-gateway
```

Comment:
This seems to conflict with @howardjohn's KEP that proposes introducing an AttachTo ObjectRef, https://github.com/kubernetes-sigs/gateway-api/pull/1868. It seems like these KEPs are intended to work together, which approach is intended here?

Reply:
From John's GEP
> See GEP-1713 for details, which this depends on.

My assumption is John will update his GEP according to the result of this one.  I'm open to discussion on the mechanics here.

My motivation for not putting it under a `infrastructure` field is because I view merging to be a logical operation done independently on the infrastructure. We're just describing a single Gateway using N resources - in order to work around the limitations we have when updating many listeners.  This is why I asked in slack whether there may be narrow use cases for certs on HTTPRoutes (https://github.com/kubernetes-sigs/gateway-api/pull/1863#discussion_r1145355857)

Reply:
The idea I was thinking of by putting this in `infrastructure` is that the split-out Gateways are then saying "I use this Gateway for my infrastructure, and Listeners I specify (along with Routes that attach to them) should be merged into that Gateway's infrastructure."

In this model, you'd have a single Gateway A that represents the actual infrastructure, and then other Gateways that reference that Gateway as their infrastructure provider. All Gateways that select Gateway A as their infrastructure provider would then be merged.

I think it achieves the same goal, without needing constructs like a boolean field to enable merging.

Reply:
OK I'm convinced.

Are we ok with `infrastructure.attachTo` or prefer a more explicit `infrastructure.parentGateway` property ?

Reply:
@howardjohn is there a reason why you used a `LocalObjectReference` vs a gateway name?

Reply:
Thanks for clarifying - I'm good with `attachTo => Gateway` mechanic. 

I just want to confirm with everyone that `infrastructure` and this specific use of `attachTo` is something that we could see in 'Core' long term? And that they're not just specific to in-cluster deployments.

Reply:
I cannot speak for everyone but my intent is that infrastructure goes beyond "in cluster"

Reply:
> I just want to confirm with everyone that `infrastructure` and this specific use of `attachTo` is something that we could see in 'Core' long term? And that they're not just specific to in-cluster deployments.

Yes, I can see this in Core in the long term. It will need to start in Extended until we are confident that it's supported everywhere though.